### ISSUE TITLE
MockServerClient reset 4 secs overhead

### ISSUE THREAD

**Issue by mbjelac** [jamesdbloom/mockserver#382]
This is a continuation of troubleshooting a problem which I originally reported in #380.

You were right, it has nothing to do with Docker. I managed to narrow it down to the `MockServerClient`.

I wrote two tests, both calling `reset`:
- one using `MockServerClient`
- one using Apache HTTP client

I monitored the traffic with Wireshark and the problem is clear: `MockServerClient` waits 4.5 seconds after the last `ACK` from the server to send the HTTP PUT request.

This can clearly be seen in the PCAP. Compare with PCAP of traffic when using Apache HTTP client.
[Download PCAPs in ZIP](https://github.com/jamesdbloom/mockserver/files/1481524/mockserver_issue_380_pcaps_1.zip)

The test I used to get the traffic is in [this gist](https://gist.github.com/mbjelac/84a9310a7c96ba6c1d802f5c2dddddc9).

The set-expectation call and others probably suffer from the same problem.

**Comment by jamesdbloom** [comment]
Ok I'll take a look into that

**Comment by jamesdbloom** [comment]
The overhead is not actually really caused by the MockServerClient specifically.

The real cause if the delay you are experiencing is the JVM slow start time.  

The `NettyHttpClient.sendRequest` timings are as follows
1: 735ms,
2: 11ms
3: 11ms
4-14: 7ms

The `NettyHttpClient` constructor timings are as follows:
1: 1479ms
2: 0
3: 0
...

However if you add the following line into the JUnit class, so that the `LoggerFactory` is loaded:

```java
private final Logger logger = LoggerFactory.getLogger(this.getClass());
```

The `NettyHttpClient` constructor timings are as follows:
1: 8ms
2: 0
3: 0
...

So what this shows is that most of the time is spent the first time you call `LoggerFactory.getLogger` in the JVM.  The remaining major chunk of time is spent the first time you call `NettyHttpClient.sendRequest`.  This is most likely due to classloading and the JIT not compiling freshly loaded bytecode until it has been executed enough times.

The fact that request calls, including both client and server time drop to 7ms (on the same machine) strongly suggests that there is nothing fundamentally wrong with the performance.  The real cause of any performance "issue" is the well known JVM slow start caused by classloading and the delay before the JIT fully optimises (in two stages) the bytecode.

I'm going to close this issue as I can't reproduce any performance impact (except for JVM slow start).  If you have additional information for investigation please re-open.

**Comment by mbjelac** [comment]
I expanded my test - see [the gist](https://gist.github.com/mbjelac/2d39cd916dae5a08836f67df4ff172be).
First I triplicated the tests and moved `MockServerClient` to a static field - it is instantiated only once. Then I added tests using only your `NettyHttpClient`, also triplicate.
Also, I added a `Logger` like you suggested.

Please run the tests. The lag is evident without checking out TCP traffic. (However, I *did* check the TCP traffic - `NettyHttpClient` tests behave exactly like `MockServerClient` tests.)

See the attached test duration screenshot. The first test is using `NettyHttpClient` and takes 1 second longer than the other two `NettyHttpClient` tests. *This* I can account for class loading. Its also evident with Apache tests - the first one takes longer. However, the first of the `MockServerClient` tests doesn't take longer because it's using the same classes as the test with the `NettyHttpClient`.

**Comment by mbjelac** [comment]
@jamesdbloom - did you get the chance to check my new info out? I cannot re-open the issue since I didn't close it (and I'm not a collaborator).

**Comment by jamesdbloom** [comment]
Thanks, I'll take a look, today or tomorrow

**Comment by jamesdbloom** [comment]
I still can't reproduce your findings, but perhaps thats because I'm using the latest version of MockServer.  I can see that because your using `OutboundHttpRequest` you must be using an old version of MockServer.

I have run the test two ways.

1. `org.mockserver.cli.Main -Dmockserver.logLevel=INFO -serverPort 1080 -proxyPort 1090` with version `5.2.4-SNAPSHOT`

<img width="399" alt="screen shot 2017-12-22 at 09 01 55" src="https://user-images.githubusercontent.com/733179/34292111-4d043a70-e6f7-11e7-97ff-93b25196bca6.png">

2. `docker run -d -p 1080:1080 -p 1090:1090 jamesdbloom/mockserver:mockserver-5.2.3`

<img width="398" alt="screen shot 2017-12-22 at 09 02 28" src="https://user-images.githubusercontent.com/733179/34292124-5bf478e2-e6f7-11e7-83ec-02cc2cc4c6eb.png">

**Comment by mbjelac** [comment]
Thanks @jamesdbloom - issue solved with latest version (5.2.3)!
I can't believe I was two major versions behind...

Apache is still approx. 5 times faster than Netty in my tests - but both are under 100 ms so that's OK. :)

Sorry for the bother, I must remember to update the version when troubleshooting.
